File: library/src/main/java/wannabe/grid/iterators/HiddenRemovalIterator.java
package wannabe.grid.iterators;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import wannabe.Voxel;
import wannabe.grid.Grid;

public final class HiddenRemovalIterator implements Iterator<Voxel> {
  private final Iterator<Voxel> realIterator;

  public HiddenRemovalIterator(Grid source) {
    List<Voxel> nonHidden = new ArrayList<>(source.size());
    for (Voxel voxel : source) {
      if (!source.neighbors(voxel).isSurrounded()) {
        nonHidden.add(voxel);
      }
    }
    realIterator = nonHidden.iterator();
  }

  @Override public boolean hasNext() {
    return realIterator.hasNext();
  }

  @Override public Voxel next() {
    return realIterator.next();
  }

  @Override public void remove() {
    // TODO should I support remove?
    realIterator.remove();
  }
}

File: library/src/main/java/wannabe/grid/iterators/HiddenRemovalIterator.kt
package wannabe.grid.iterators

import wannabe.Voxel
import wannabe.grid.Grid
import java.util.ArrayList

// TODO this should be a streaming solution not a caching one
class HiddenRemovalIterator(
  source: Grid
) : MutableIterator<Voxel> {
  private val realIterator: MutableIterator<Voxel>

  init {
    val nonHidden = source.filter { !source.neighbors(it).isSurrounded }
        .toMutableList()
    realIterator = nonHidden.iterator()
  }

  override fun hasNext(): Boolean = realIterator.hasNext()

  override fun next(): Voxel = realIterator.next()

  // TODO should I support remove?
  override fun remove() = realIterator.remove()
}
